File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.js
// DFS이용
function solution(numbers, target) {
  let cnt = 0
  let signs = ['+', '-']

  function dfs(level, sum, target) {
    if (level === numbers.length) {
      if (sum === target) cnt ++
      return
    }

    for (let sign of signs) {
      dfs(level+1, sum+Number(sign+numbers[level]),target)
    }
  }

  dfs(0,0,target)

  return cnt
}

console.log(solution([1, 1, 1, 1, 1],3))

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# dfs로. bfs도 되는건가??
# 생각해보니까, 트리형식으로 2개로 나눠서 내려가는거니까 bfs같기도?(1,-1)
answer = 0


def dfs(arr, level, sum, target):
    global answer
    if level == len(arr):
        if sum == target:
            answer += 1
        return

    for ele in [1, -1]:
        dfs(arr, level + 1, sum + (arr[level]) * ele, target)


def solution(numbers, target):
    global answer
    dfs(numbers, 0, 0, target)
    return answer


# print(solution([1, 1, 1, 1, 1], 3))
# print(solution([4, 1, 2, 1], 4))
